relayer/provider/cosmos/tx.go

```diff
@@ -1204,6 +1204,269 @@ func (cc *CosmosProvider) MsgTimeoutOnClose(msgTransfer provider.PacketInfo, pro
+		ProofClient:     proof.ClientStateProof,
+		ProofConsensus:  proof.ConsensusStateProof,
+		ConsensusHeight: proof.ClientState.GetLatestHeight().(clienttypes.Height),
+		Signer:          signer,
+	}
+
+	return NewCosmosMessage(msg), nil
+}
+
+func (cc *CosmosProvider) ConnectionProof(ctx context.Context, msgOpenAck provider.ConnectionInfo, latest provider.LatestBlock) (provider.ConnectionProof, error) {
+	connState, err := cc.QueryConnection(ctx, int64(latest.Height), msgOpenAck.ConnID)
+	if err != nil {
+		return provider.ConnectionProof{}, err
+	}
+
+	return provider.ConnectionProof{
+		ConnectionStateProof: connState.Proof,
+		ProofHeight:          connState.ProofHeight,
+	}, nil
+}
+
+func (cc *CosmosProvider) MsgConnectionOpenConfirm(msgOpenAck provider.ConnectionInfo, proof provider.ConnectionProof) (provider.RelayerMessage, error) {
+	signer, err := cc.Address()
+	if err != nil {
+		return nil, err
+	}
+	msg := &conntypes.MsgConnectionOpenConfirm{
+		ConnectionId: msgOpenAck.CounterpartyConnID,
+		ProofAck:     proof.ConnectionStateProof,
+		ProofHeight:  proof.ProofHeight,
+		Signer:       signer,
+	}
+
+	return NewCosmosMessage(msg), nil
+}
+
+func (cc *CosmosProvider) MsgChannelOpenInit(info provider.ChannelInfo, proof provider.ChannelProof) (provider.RelayerMessage, error) {
+	signer, err := cc.Address()
+	if err != nil {
+		return nil, err
+	}
+	msg := &chantypes.MsgChannelOpenInit{
+		PortId: info.PortID,
+		Channel: chantypes.Channel{
+			State:    chantypes.INIT,
+			Ordering: info.Order,
+			Counterparty: chantypes.Counterparty{
+				PortId:    info.CounterpartyPortID,
+				ChannelId: "",
+			},
+			ConnectionHops: []string{info.ConnID},
+			Version:        info.Version,
+		},
+		Signer: signer,
+	}
+
+	return NewCosmosMessage(msg), nil
+}
+
+func (cc *CosmosProvider) ChannelProof(ctx context.Context, msg provider.ChannelInfo, latest provider.LatestBlock) (provider.ChannelProof, error) {
+	channelRes, err := cc.QueryChannel(ctx, int64(latest.Height), msg.ChannelID, msg.PortID)
+	if err != nil {
+		return provider.ChannelProof{}, err
+	}
+	return provider.ChannelProof{
+		Proof:       channelRes.Proof,
+		ProofHeight: channelRes.ProofHeight,
+		Version:     channelRes.Channel.Version,
+		Ordering:    channelRes.Channel.Ordering,
+	}, nil
+}
+
+func (cc *CosmosProvider) MsgChannelOpenTry(msgOpenInit provider.ChannelInfo, proof provider.ChannelProof) (provider.RelayerMessage, error) {
+	signer, err := cc.Address()
+	if err != nil {
+		return nil, err
+	}
+	msg := &chantypes.MsgChannelOpenTry{
+		PortId:            msgOpenInit.CounterpartyPortID,
+		PreviousChannelId: msgOpenInit.CounterpartyChannelID,
+		Channel: chantypes.Channel{
+			State:    chantypes.TRYOPEN,
+			Ordering: proof.Ordering,
+			Counterparty: chantypes.Counterparty{
+				PortId:    msgOpenInit.PortID,
+				ChannelId: msgOpenInit.ChannelID,
+			},
+			ConnectionHops: []string{msgOpenInit.CounterpartyConnID},
+			// TODO wire this up
+			Version: "ics20-1",
```

Comment:
@jtieri should the version always be the same for the counterparty? If so, we can use `proof.Version` here also.

Reply:
Thanks for the info. Maybe in the future we would need to add a flag for the counterparty version also and wire that through the the `PathProcessor`, but seems reasonable enough for now for them to be the same. I changed it to be the same and added a comment with this rationale to replace the TODO.